src/storage/compression/zstd.cpp

```diff
@@ -0,0 +1,1129 @@
+
+		idx_t page_size = block_manager.GetBlockSize() - sizeof(block_id_t);
+		idx_t remaining_compressed_data = scan_state.metadata.compressed_size - scan_state.compressed_scan_count;
+		scan_state.in_buffer.size = MinValue<idx_t>(page_size, remaining_compressed_data);
+	}
+
+	void DecompressString(ZSTDVectorScanState &scan_state, data_ptr_t destination, idx_t uncompressed_length) {
+		if (uncompressed_length == 0) {
+			return;
+		}
+
+		duckdb_zstd::ZSTD_outBuffer out_buffer;
+
+		out_buffer.dst = destination;
+		out_buffer.pos = 0;
+		out_buffer.size = uncompressed_length;
+
+		while (true) {
+			idx_t old_pos = scan_state.in_buffer.pos;
+			size_t res = duckdb_zstd::ZSTD_decompressStream(
+			    /* zds = */ decompression_context,
+			    /* output =*/&out_buffer,
+			    /* input =*/&scan_state.in_buffer);
+			scan_state.compressed_scan_count += scan_state.in_buffer.pos - old_pos;
+			if (duckdb_zstd::ZSTD_isError(res)) {
+				throw InvalidInputException("ZSTD Decompression failed: %s", duckdb_zstd::ZSTD_getErrorName(res));
+			}
+			if (out_buffer.pos == out_buffer.size) {
+				break;
+			}
+			// Did not fully decompress, it needs a new page to read from
+			LoadNextPageForVector(scan_state);
+		}
+	}
+
+	void Skip(ZSTDVectorScanState &scan_state, idx_t count) {
+		if (!skip_buffer) {
+			skip_buffer = Allocator::DefaultAllocator().Allocate(duckdb_zstd::ZSTD_DStreamOutSize());
+		}
+
+		D_ASSERT(scan_state.scanned_count + count <= scan_state.metadata.count);
+
+		// Figure out how much we need to skip
+		string_length_t *string_lengths = &scan_state.string_lengths[scan_state.scanned_count];
+		idx_t uncompressed_length = 0;
+		for (idx_t i = 0; i < count; i++) {
+			uncompressed_length += string_lengths[i];
+		}
+
+		// Skip that many bytes by decompressing into the skip_buffer
+		idx_t remaining = uncompressed_length;
+		while (remaining) {
+			idx_t to_scan = MinValue<idx_t>(skip_buffer.GetSize(), remaining);
+			DecompressString(scan_state, skip_buffer.get(), to_scan);
+			remaining -= to_scan;
+		}
+		scan_state.scanned_count += count;
+		scanned_count += count;
+	}
+
+	void ScanInternal(ZSTDVectorScanState &scan_state, idx_t count, Vector &result, idx_t result_offset) {
+		D_ASSERT(scan_state.scanned_count + count <= scan_state.metadata.count);
+		D_ASSERT(result.GetType().id() == LogicalTypeId::VARCHAR);
+
+		string_length_t *string_lengths = &scan_state.string_lengths[scan_state.scanned_count];
+		idx_t uncompressed_length = 0;
+		for (idx_t i = 0; i < count; i++) {
+			uncompressed_length += string_lengths[i];
+		}
+		auto empty_string = StringVector::EmptyString(result, uncompressed_length);
+		auto uncompressed_data = empty_string.GetDataWriteable();
+		auto string_data = FlatVector::GetData<string_t>(result);
+		for (idx_t i = 0; i < count; i++) {
+			data_ptr_t start_of_uncompressed_string = reinterpret_cast<data_ptr_t>(uncompressed_data);
+			DecompressString(scan_state, start_of_uncompressed_string, string_lengths[i]);
```

Comment:
Can we unify the decompress string calls across multiple strings? That should show some perf improvements on decompression